third_party/move/move-vm/runtime/src/runtime_type_checks_async.rs

```diff
@@ -0,0 +1,785 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Defines async type checker that abstractly interprets Move bytecode to perform type checks
+//! based on the execution trace.
+//!
+//! The type checker should ideally be run in parallel and as a post-execution hook. Otherwise,
+//! if there is no parallelism or not enough transactions, running checks in-place is preferred.
+//!
+//! The type checker can safely use [UnmeteredGasMeter], or other unmetered APIs (e.g., fpr module
+//! loading) because trace records only successful execution of instructions, and so the gas must
+//! have been charged during execution.
+//!
+//! The type checker is also not expected to fail. Any type check violations must be caught by the
+//! bytecode verifier, so the runtime checks are an additional safety net. Because of this property
+//! it is safe to run these checks after the actual execution.
+
+use crate::{
+    config::VMConfig,
+    execution_tracing::Trace,
+    frame::Frame,
+    frame_type_cache::{AllRuntimeCaches, FrameTypeCache, PerInstructionCache},
+    interpreter::{CallStack, Stack},
+    interpreter_caches::InterpreterFunctionCaches,
+    loader::FunctionHandle,
+    reentrancy_checker::CallType,
+    runtime_type_checks::{
+        verify_pack_closure, FullRuntimeTypeCheck, RuntimeTypeCheck, UntrustedOnlyRuntimeTypeCheck,
+    },
+    LoadedFunction, LoadedFunctionOwner, ModuleStorage,
+};
+use move_binary_format::{
+    errors::{Location, PartialVMError, PartialVMResult, VMResult},
+    file_format::{Bytecode, FunctionHandleIndex, FunctionInstantiationIndex},
+};
+use move_core_types::function::ClosureMask;
+use move_vm_types::{
+    gas::UnmeteredGasMeter,
+    loaded_data::runtime_types::{Type, TypeBuilder},
+    ty_interner::{InternedTypePool, TypeVecId},
+    values::{Locals, Value},
+};
+use std::{
+    cell::RefCell,
+    collections::{btree_map::Entry, VecDeque},
+    iter,
+    rc::Rc,
+};
+
+/// This ensures error location is set in the same way as in the interpreter.
+macro_rules! set_err_info {
+    ($frame:ident, $e:expr) => {{
+        $e.at_code_offset($frame.function.index(), $frame.pc)
+            .finish($frame.location())
+    }};
+}
+
+/// Exit codes returned when type frame reaches a control-flow instructions like calls, etc.
+enum ExitCode {
+    /// Replay is done. No more instructions need to be replayed.
+    Done,
+    /// Function returns the control to the caller.
+    Return,
+    /// Function statically calls into a non-generic function.
+    Call(FunctionHandleIndex),
+    /// Function statically calls into a generic function.
+    CallGeneric(FunctionInstantiationIndex),
+    /// Function dynamically calls a closure (must be recorded in the trace).
+    CallClosure,
+}
+
+/// Runtime type checker based on tracing.
+pub struct AsyncRuntimeTypeCheck<'a, T> {
+    /// Stores type information for type checks.
+    stack: Stack,
```

Comment:
`type_stack`?  Also is the data structure from the interpreter?

Reply:
I tried to re-use as much as possible, we can probably have a custom stack here (and better have a custom stack), but for now this makes behavior and APIs identical for async vs in-place modes